pkg/controllers/management/auth/user_attribute_handler.go

```diff
@@ -28,6 +28,10 @@ func (ua *UserAttributeController) sync(key string, obj *v3.UserAttribute) (runt
 		return nil, nil
 	}
 
+	if obj.ExtraByProvider == nil {
```

Comment:
If I understand correctly, what you're proposing is already how it works and is part of the problem. When a user re-logs in to the UI and calls [NewLoginToken](https://github.com/rancher/rancher/blob/13482caab5fb28b9445494d95f865eeb5d39c44e/pkg/auth/tokens/manager.go#L648) their user attributes are refreshed based on the data in the token. The problem is that this is the only time this happens. So if the user has never logged in to the UI since 2.6.2 then their attributes aren't refreshed. This just forces the startup logic to bring the user's attributes into sync with what they would be if they had gone through the UI login flow.

Reply:
Sorry for the confusion - I should have linked the relevant code in my explanation. I wasn't referring to the login logic but the [Authenticate](https://github.com/rancher/rancher/blob/13482caab5fb28b9445494d95f865eeb5d39c44e/pkg/auth/requests/authenticate.go#L156-L160) logic, which is what takes the token from a request and retrieves the information for that user. The issue is that rather than using the extras specified on the UserAttribute, it [makes the call to the provider](https://github.com/rancher/rancher/blob/13482caab5fb28b9445494d95f865eeb5d39c44e/pkg/auth/requests/authenticate.go#L185) to get the extras. 

This is what creates issue - the UserAttributes/impersonation rules have outdated extras compared to the user. This new logic (where we refresh the extras on attribute sync) could lead to situations where a user issued a request, Rancher found new extras, but those extras aren't added to the attributes until refresh (though as said above, since our extras are static, not likely today). I see two solutions to ensure that `Authenticate`'s extras always agree with `UserAttribute`'s extras:

- In `Authenticate`, don't call to the provider to get extras, just use what is on the `UserAttribute`. Since this will get updated on refresh, it will be eventually accurate, but this way we won't have impersonation issues.
- Update the `UserAttribute` with the new extras when doing `Authenticate`. This way the `UserAttribute` (and therefore impersonation) will always have the latest info.

I think that even with these other items the logic to update extras on refresh is necessary and should stay.

Reply:
Resolving this discussion, because:

- The current "reach out to the provider" logic is only called when there aren't extras specified on the UserAttribute. This PR aims to cause an Attribute refresh on boot, so in theory, this should never occur past this point. While removing this would ensure that this operates as expected, removing code like this can cause unexpected issues. In short - the problem code is unlikely to be reached with this fix, and removing it is not worth the risk that removing such code poses.

Reply:
@MbolotSuse All the ucecases you have mentioned are already handled.
- In `Authenticate` - The logic is to use the extras specified on the UserAttribute. If the extras do not contain principalid/displayname only then is the provider called which uses the userPrincipal from the request Token. Also if request token fails to get that info, final backup is data from User object of the request which will be updated info.